37626 remove references to the ads from heliumanalyserefficiency algorithm

yusufjimoh opened the pull request:
### Description of work
Remove References to ADS from HeliumAnalyserEfficiencyAlgorithm

#### Summary of work
- Remove references from ADS, which previously handled validation and creation of the input workspace group. Now, the input workspace group from the GUI is validated and used directly in the algorithm.
- Adjust failing unit tests that relied on ADS to ensure they pass successfully.
- Add a test to ensure the algorithm works successfully when used as a child algorithm.

Fixes #37626 

### To test:
- Execute the HeliumAnalyserEfficiencyAlgorithm 
- Algorithm should execute successfully
- All Tests cases should pass 

This does not require release notes because it involves internal code changes that have no effect on the current functionality of the algorithm

yusufjimoh committed:
after removing the reference from ADS from the code, the failing unit tests were modified

yusufjimoh committed:
Clean Up Code in HeliumAnalyserEfficiency.cpp

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -51,9 +51,9 @@ void HeliumAnalyserEfficiency::init() {
-  declareProperty(
-      std::make_unique<WorkspaceProperty<>>(PropertyNames::INPUT_WORKSPACE, "", Direction::Input, validator),
-      "Input group workspace to use for polarization calculation");
+  declareProperty(std::make_unique<WorkspaceProperty<Mantid::API::WorkspaceGroup>>(PropertyNames::INPUT_WORKSPACE, "",
```
Apologies, I can't get a comment onto the relevant line as it's not part of the code that has been changed, but lines 120-137 (the methods `validateGroupInput` and `processGroups`) can be removed now that we're setting the input workspace property type to a `WorkspaceGroup`.

By default, when a Mantid algorithm has a normal workspace input property (rather than a `WorkspaceGroup` input property) then passing a workspace group to this will loop through the group and call the algorithm against each child workspace individually. The `processGroups` method is used as a way of overriding this default behaviour when we want the algorithm to do something different with a workspace group. We needed to use this method when the algorithm was set up as it was originally, however now that we're explicitly specifying the input property to be a `WorkspaceGroup` it will work as we need it to without defining the `processGroups` method.

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -51,9 +51,9 @@ void HeliumAnalyserEfficiency::init() {
   auto validator = std::make_shared<CompositeValidator>();
```
I don't think we're using this validator any more (it doesn't seem to work for WorkspaceGroup properties) so these three lines of code can be removed. We do, however, still need this validation that checks that the input workspace is in Wavelength units, as when I tested I was able to pass in a workspace in TOF units which shouldn't be possible. I would recommend taking a look at the workspace unit validation in either the `FlipperEfficiency` or `PolarizerEfficiency` (recently merged in) algorithms as we will need to add something similar here to ensure the validation takes place.

There is a unit test that checks for this, `testNonWavelengthInput`, however it's not checking the specific error message that's being thrown so I suspect there is some other reason why the algorithm is throwing an error here that is unrelated to the workspace units. I would recommend starting by re-working that test so that it's failing with the code changes here, and then look at adding the validation in to get the test to pass.

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -103,20 +103,15 @@ void HeliumAnalyserEfficiency::init() {
-  if (!AnalysisDataService::Instance().doesExist(inputWorkspaceName)) {
-    errorList[PropertyNames::INPUT_WORKSPACE] =
-        "The input workspace " + inputWorkspaceName + " does not exist in the ADS.";
+  WorkspaceGroup_sptr ws = getProperty(PropertyNames::INPUT_WORKSPACE);
```
I think we can set this to be a `const` variable to highlight that we're not planning to make changes to the `ws` variable.

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -103,20 +103,15 @@ void HeliumAnalyserEfficiency::init() {
-        "The input workspace " + inputWorkspaceName + " does not exist in the ADS.";
+  WorkspaceGroup_sptr ws = getProperty(PropertyNames::INPUT_WORKSPACE);
+
+  if (!ws->isGroup()) {
```
The `WorkspaceGroup` input seems to offer a lot of built in validation so this doesn't actually cause an issue in practice, but in many cases when retrieving property values the code on line 106 would return a `nullptr` if something was passed in that couldn't be converted to a `WorkspaceGroup_sptr`. So the check we want to put here is: `if(ws == nullptr)`.

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -103,20 +103,15 @@ void HeliumAnalyserEfficiency::init() {
-            "The input group workspace must have four periods corresponding to the four spin configurations.";
-      }
+    const auto wsGroup = std::dynamic_pointer_cast<WorkspaceGroup>(ws);
+    if (wsGroup->size() != 4) {
```
Now that we're getting the workspace group directly from the property on line 106, this can be streamlined to remove the dynamic_pointer_cast (i.e. remove line 111) and just have `if(ws->size() != 4)`. To improve readability of the code, we might like to name the original variable on line 106 `wsGroup` so that we can continue to use that name here in the `if` statement and it's clear to future developers that this variable holds a workspace group.

rbauststfc commented on Framework/Algorithms/test/PolarizationCorrections/HeliumAnalyserEfficiencyTest.h:
```diff
@@ -42,7 +42,7 @@ class HeliumAnalyserEfficiencyTest : public CxxTest::TestSuite {
     heliumAnalyserEfficiency->initialize();
-    heliumAnalyserEfficiency->setProperty("InputWorkspace", ws1->getName());
+    TS_ASSERT_THROWS(heliumAnalyserEfficiency->setProperty("InputWorkspace", ws1), std::invalid_argument &);
     heliumAnalyserEfficiency->setProperty("OutputWorkspace", "P");
```
Can we remove the two lines after this, as I'm not sure they will be reached if the test has already thrown by this point?

rbauststfc replied:
Looking at this again, I think the two lines after this assert are actually reached. I think it's probably sufficient for the purpose of the test just to have the first assert in there, but I don't think it's an issue to keep the extra lines too if preferred, so feel free to go with what you think on this.

rbauststfc reviewed (commented):
Thanks very much for this, a good start on a change that will really streamline the code here, and good to see a new unit test to check we can run the algorithm successfully as a child algorithm.

There are a few bits I've noticed that I've added comments about below. In particular we will need to add the workspace unit validation back in and make sure we have an effective unit test that picks up when that validation isn't happening. There are also some methods that can be removed now that we're using a WorkspaceGroup property (because of some very Mantid-specific algorithm behaviour that I've tried to explain below 🙂).

Please could we also make sure to add the 6.11 milestone to this PR, as well as the LSS and SANS labels.

yusufjimoh committed:
Update Validation for InputWorkspace Group

- Added Validation to ensure all input workspaces have units of wavelength
- Apply Suggested Chnanges from PR

yusufjimoh commented:
Thank you for your review and contextual explanation of the code changes that needed to be made, which was really helpful. The PR has been updated, notably with the addition of validation to ensure that input workspaces have units of wavelengths.

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -96,56 +95,42 @@ void HeliumAnalyserEfficiency::init() {
   // First we extract the individual workspaces corresponding to each spin configuration from the group workspace
-  const auto groupWorkspace =
-      AnalysisDataService::Instance().retrieveWS<WorkspaceGroup>(getProperty(PropertyNames::INPUT_WORKSPACE));
+  WorkspaceGroup_sptr groupWorkspace = getProperty(PropertyNames::INPUT_WORKSPACE);
```
Sorry, I didn't spot this the first time through, but this looks like the variable used to be marked as `const` to indicate that it isn't going to be changed. Please could we put the `const` back here? As we seem to be using "west const" in this file, we would define the variable as `const WorkspaceGroup_sptr groupWorkspace = `

rbauststfc commented on Framework/Algorithms/src/PolarizationCorrections/HeliumAnalyserEfficiency.cpp:
```diff
@@ -96,56 +95,42 @@ void HeliumAnalyserEfficiency::init() {
-            "The input group workspace must have four periods corresponding to the four spin configurations.";
-      }
+    for (size_t i = 0; i < wsGroup->size(); ++i) {
+      MatrixWorkspace_sptr const stateWs = std::dynamic_pointer_cast<MatrixWorkspace>(wsGroup->getItem(i));
```
This is very much an edge case that is unlikely to be encountered in practice, but I can cause a crash by running this script:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

wsTPara = CreateEmptyTableWorkspace()
wsTPara1 = CreateEmptyTableWorkspace()
wsTAnti = CreateEmptyTableWorkspace()
wsTAnti1 = CreateEmptyTableWorkspace()

grp = GroupWorkspaces([wsTPara,wsTAnti,wsTPara1,wsTAnti1])
eff = HeliumAnalyserEfficiency(grp, SpinStates='11,01,00,10')
```

This creates an input group of table workspaces, rather than the expected input group of matrix workspaces. On line 122 of the algorithm where we are getting the individual workspace from the group and casting it to a `MatrixWorkspace`, the variable `stateWs` will be a `nullptr` if the type of the workspace is not a `MatrixWorkspace`. When we use this later inside `validateInputWorkspace` we then get a crash when we try to get the axis from it on line 99). To avoid the crash we should add a line to the start of the `validateInputWorkspace` method that checks for a `nullptr` and returns straight away (with appropriate error message) if one is found. So the `validateInputWorkspace` method would be refactored to be:

```
void validateInputWorkspace(MatrixWorkspace_sptr const &workspace, std::map<std::string, std::string> &errorList) {
  if (workspace == nullptr) {
    errorList[PropertyNames::INPUT_WORKSPACE] = "All input workspaces must be of type MatrixWorkspace.";
    return;
  }

  Kernel::Unit_const_sptr unit = workspace->getAxis(0)->unit();
  if (unit->unitID() != "Wavelength") {
    errorList[PropertyNames::INPUT_WORKSPACE] = "All input workspaces must be in units of Wavelength.";
  }
}
```

rbauststfc replied:
This is very superficial, but as there are other changes to make on this PR, could we move the use of `const` here to the left of the variable declaration ("west const"), to be consistent with the rest of the file? So this would become: `const MatrixWorkspace_sptr stateWs = std::dynamic_pointer_cast<MatrixWorkspace>(wsGroup->getItem(i));`

rbauststfc reviewed (commented):
Thanks very much for the edits so far, this is really starting to take shape. There are a couple of things I noticed on re-review, two very minor and one crash (edge case, but still worth fixing). Once those are sorted then this will be looking good.

yusufjimoh committed:
Update Validation to Ensure Input Workspaces are of type MatrixWorkspace

rbauststfc reviewed (approved):
Thanks for this, the code changes are all looking really good now. We're making good use of the workspace group property rather than relying on the ADS. The workspace unit validation has been re-instated successfully with a working test and the code is now robust to an input workspace group with child workspaces of an incorrect type. This is looking good to be picked up by a gatekeeper now 👍
